test/e2e/all_in_one_test.go

```diff
@@ -295,10 +354,47 @@ func verifyIngress(ctx context.Context, t *testing.T, env environments.Environme
 	t.Log("finding the kong proxy service ip")
 	svc, err := env.Cluster().Client().CoreV1().Services(namespace).Get(ctx, "kong-proxy", metav1.GetOptions{})
 	require.NoError(t, err)
-	require.Len(t, svc.Status.LoadBalancer.Ingress, 1)
-	proxyIP := svc.Status.LoadBalancer.Ingress[0].IP
+	proxyIP := ""
+	require.NotEqual(t, svc.Spec.Type, svc.Spec.ClusterIP)
+	if svc.Spec.Type == corev1.ServiceTypeLoadBalancer {
+		if len(svc.Status.LoadBalancer.Ingress) > 0 {
+			proxyIP = svc.Status.LoadBalancer.Ingress[0].IP
+		}
+	}
+	// the above failed to find an address. either the LB didn't provision or we're using a NodePort
+	if proxyIP == "" {
+		var port int32
+		for _, sport := range svc.Spec.Ports {
+			if sport.Name == "kong-proxy" || sport.Name == "proxy" {
+				port = sport.NodePort
+			}
+		}
+		var extAddrs []string
+		var intAddrs []string
+		nodes, err := env.Cluster().Client().CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+		require.NoError(t, err)
+		for _, node := range nodes.Items {
+			for _, naddr := range node.Status.Addresses {
+				if naddr.Type == corev1.NodeExternalIP {
+					extAddrs = append(extAddrs, naddr.Address)
+				}
+				if naddr.Type == corev1.NodeInternalIP {
+					extAddrs = append(intAddrs, naddr.Address)
+				}
+			}
+		}
+		require.NotZero(t, len(extAddrs)+len(intAddrs))
+		// local clusters (KIND, minikube) typically provide no external addresses, but their internal addresses are
+		// routeable from their host. We prefer external addresses if they're available, but fall back to internal
+		// in their absence
+		if len(extAddrs) > 0 {
+			proxyIP = fmt.Sprintf("%v:%v", extAddrs[0], port)
+		} else {
+			proxyIP = fmt.Sprintf("%v:%v", intAddrs[0], port)
+		}
```

Comment:
straighten out the flow to reduce surprises

```suggestion
		// local clusters (KIND, minikube) typically provide no external addresses, but their internal addresses are
		// routeable from their host. We prefer external addresses if they're available, but fall back to internal
		// in their absence
		if len(extAddrs) > 0 {
			proxyIP = fmt.Sprintf("%v:%v", extAddrs[0], port)
		} else if len(intAddrs) > 0 {
			proxyIP = fmt.Sprintf("%v:%v", intAddrs[0], port)
		} else {
			assert.Fail(t, "both extAddrs and intAddrs are empty")
		}
```

Reply:
IIRC from somewhere nodes _must_ have at least one internal address, but can't find that reference and the additional case can't hurt anything :shrug: